pkg/ruler/ruler_test.go

```diff
@@ -844,9 +899,89 @@ func TestGetRules(t *testing.T) {
 				time.Sleep(100 * time.Millisecond)
 			}
 
+			scrape := func(rID string, scrapeCh chan prometheus.Metric, doneCh chan struct{}) {
+				go func() {
+					tick := time.NewTicker(10 * time.Millisecond)
+					for {
+						select {
+						case <-tick.C:
+							if len(scrapeCh) >= 100 {
+								continue
+							}
+							m := rulerManagerMap[rID]
+							data := m.userManagerMetrics.regs.BuildMetricFamiliesPerUser()
+							data.SendSumOfCountersPerUserWithLabels(scrapeCh, m.userManagerMetrics.IterationsScheduled, "prometheus_rule_group_iterations_total", "rule_group")
+						case <-doneCh:
+							return
+						}
+					}
+				}()
+			}
+
+			waitUntil := func(rID string, user string, ruleGroupList rulespb.RuleGroupList) {
+				scrapeCh := make(chan prometheus.Metric, 100)
+				doneCh := make(chan struct{})
+				scrape(rID, scrapeCh, doneCh)
+				m := map[string]float64{}
+				for _, rgDesc := range ruleGroupList {
+					ruler := rulerAddrMap[rID]
+					if tc.sharding {
+						subRing := ruler.ring.ShuffleShard(user, tc.shuffleShardSize)
+						owns, err := instanceOwnsRuleGroup(subRing, rgDesc, validation.DisabledRuleGroups{}, ruler.lifecycler.GetInstanceAddr(), false)
```

Comment:
it feels that during error the waitUntil should also error instead of continuing the test. There are no disabled rule groups so you can only get a real error.